src/libraries/System.Private.CoreLib/src/System/ArgumentOutOfRangeException.cs

```diff
@@ -85,5 +88,121 @@ public override string Message
 
         // Gets the value of the argument that caused the exception.
         public virtual object? ActualValue => _actualValue;
+
+        [DoesNotReturn]
+        private static void ThrowZero(string? paramName)
+        {
+            throw new ArgumentOutOfRangeException(paramName, SR.Format(SR.ArgumentOutOfRange_Generic_MustBeNonZero, paramName));
+        }
+
+        [DoesNotReturn]
+        private static void ThrowNegative(string? paramName)
+        {
+            throw new ArgumentOutOfRangeException(paramName, SR.Format(SR.ArgumentOutOfRange_Generic_MustBeNonNegative, paramName));
+        }
+
+        [DoesNotReturn]
+        private static void ThrowNegativeOrZero(string? paramName)
+        {
+            throw new ArgumentOutOfRangeException(paramName, SR.Format(SR.ArgumentOutOfRange_Generic_MustBeNonNegativeNonZero, paramName));
+        }
+
+        [DoesNotReturn]
+        private static void ThrowGreater(string? paramName, object other)
```

Comment:
We should probably make these `object`s into `T`s.  While it's not ideal to have the extra generic methods, the primary use case is for value type Ts and having these be object will lead to code bloat at the call site with having to box.  We don't care about the actual allocation for that boxing, as it'll only happen on a code path that's about to throw, but it will increase the size of the call site, which we'd like to avoid.

Reply:
SR.Format doesn't have generic overload so I thought it's fine to have `object` here too

Reply:
It's not the body of this method that's relevant but rather the call site.  The boxing to create the `object other` happens at the call site.  If it's changed to be generic, the boxing will happen inside this method instead, which is where we want to push all of the expense.

Reply:
Thanks for the explanation, I'll change it